dtls.c

```diff
@@ -3478,39 +3506,37 @@ check_certificate_request(dtls_context_t *ctx,
 
   assert(is_key_exchange_ecdhe_ecdsa(peer->handshake_params->cipher_index));
 
-  data += DTLS_HS_LENGTH;
-
   if (data_length < DTLS_HS_LENGTH + 5) {
     dtls_alert("the packet length does not match the expected\n");
     return dtls_alert_fatal_create(DTLS_ALERT_DECODE_ERROR);
   }
 
-  i = dtls_uint8_to_int(data);
-  data += sizeof(uint8);
-  if (i + 1 > data_length) {
-    dtls_alert("the certificate types are too long\n");
-    return dtls_alert_fatal_create(DTLS_ALERT_DECODE_ERROR);
-  }
+  data += DTLS_HS_LENGTH;
+  data_length -= DTLS_HS_LENGTH;
+
+  GET_VAR_FIELD(i, data, data_length, uint8, DTLS_ALERT_DECODE_ERROR,
+                "CertificateRequest, certificate_types");
 
   auth_alg = 0;
   for (; i > 0 ; i -= sizeof(uint8)) {
     if (dtls_uint8_to_int(data) == TLS_CLIENT_CERTIFICATE_TYPE_ECDSA_SIGN
-	&& auth_alg == 0)
-      auth_alg = dtls_uint8_to_int(data);
+        && auth_alg == 0) {
+      auth_alg = TLS_CLIENT_CERTIFICATE_TYPE_ECDSA_SIGN;
+      data += i;
```

Comment:
This looks strange. We are trying to read single bytes here, correct (i.e., `sizeof(uint8_t)`)? When `i` happens to be greater than 1, we would advance `data` by i, i-1, i-2, ...

I suggest to re-arrange this loop as follows:
```c
for (; (i > 0) && (auth_alg == 0); i -= sizeof(uint8)) {
  if (dtls_uint8_to_int(data) == TLS_CLIENT_CERTIFICATE_TYPE_ECDSA_SIGN) {
    auth_alg = TLS_CLIENT_CERTIFICATE_TYPE_ECDSA_SIGN;
  }
  data += sizeof(uint8);
  data_length -= sizeof(uint8);
}
```

Reply:
> correct (i.e., sizeof(uint8_t))?

The `if` breaks the loop, because the proper value is found.
`i` contains the (left) field length, and in the case of the break, that left bytes are skipped.

Reply:
```
    if (dtls_uint8_to_int(data) == TLS_CLIENT_CERTIFICATE_TYPE_ECDSA_SIGN) {
      auth_alg = TLS_CLIENT_CERTIFICATE_TYPE_ECDSA_SIGN;
      // skip the rest of the field value
      data += i;
      data_length -=i;
      break;
    }
```